core_v14/src/metadata/version_14.rs

```diff
@@ -29,30 +29,34 @@ pub fn decode(meta: RuntimeMetadataV14) -> Result<Metadata, MetadataError> {
 
 	// Gather information about the pallets in use:
 	for pallet in meta.pallets {
-		let mut calls = HashMap::new();
-		if let Some(call_md) = pallet.calls {
-			// Get the type representing the variant of available calls:
-			let call_ty_id = call_md.ty.id();
-			let call_ty = registry.resolve(call_ty_id).ok_or(MetadataError::TypeNotFound(call_ty_id))?;
-
-			// Expect that type to be a variant:
-			let call_ty_def = call_ty.type_def();
-			let call_variant = match call_ty_def {
-				scale_info::TypeDef::Variant(variant) => variant,
-				_ => {
-					return Err(MetadataError::ExpectedVariantType { got: format!("{:?}", call_ty_def) });
-				}
-			};
+		let calls = pallet
+			.calls
+			.map(|call_md| {
+				let mut call_variant_indexes = HashMap::new();
+
+				// Get the type representing the variant of available calls:
+				let calls_type_id = call_md.ty;
+				let calls_type = registry
+					.resolve(calls_type_id.id())
+					.ok_or_else(|| MetadataError::TypeNotFound(calls_type_id.id()))?;
 
-			// Treat each variant as a function call and push to our calls list
-			for variant in call_variant.variants() {
-				// Allow case insensitive matching; lowercase the name:
-				let name = variant.name().to_ascii_lowercase();
-				let args = variant.fields().iter().map(|field| *field.ty()).collect();
+				// Expect that type to be a variant:
+				let calls_type_def = calls_type.type_def();
+				let calls_variant = match calls_type_def {
+					scale_info::TypeDef::Variant(variant) => variant,
+					_ => {
+						return Err(MetadataError::ExpectedVariantType { got: format!("{:?}", calls_type_def) });
+					}
+				};
+
+				// Store the mapping from u8 index to variant slice index or quicker decode lookup:
+				for (idx, variant) in calls_variant.variants().iter().enumerate() {
+					call_variant_indexes.insert(variant.index(), idx);
+				}
```

Comment:
Can't we use `FromIterator` here somehow, dunno, `call_variant_indexes = calls_variant.variants().iter().enumerate().collect()`?

Reply:
random thought: perhaps `call_variant_indexes` should be an `IndexMap`? I guess these collections are small enough that it doesn't matter.

Reply:
Ah, yup we can now collect straight into a `HashMap`; good call!

I looked at IndexMap briefly; it's all about preserving insertion order by the sounds of it. What was your thinking around using it?

(since the index is a u8 I could see that a `[usize; 256]` or something could be used if we wanted)

Reply:
> What was your thinking around using it?

It was a random thought, something like this: given the purpose of this hashmap is to map from one index to another index, we could sort of use insertion order (`idx` here) instead of a hashed key. So IndexMap's `IndexSet` type. Or, as you suggest, a fixed sized array.
I don't think it's a good optimization to make at this point though so, "no action required"!